### PR TITLE
Issue: MinificationMinify files along with versioning 

### PR DESCRIPTION
**Pull request by earora97**

### CODE REVIEW

File: minify.sh [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+all_js_files=("static/materialize/js/materialize.js"
+    "static/js/bloodhound.js"
+    "static/js/main.js"
+    "static/js/jquery.js"
+    "static/js/corejs-typeahead.bundle.js"
+    "static/js/calendar.js"
+    "static/js/web2py.js"
+    "static/js/web2py-bootstrap3.js"
+    #"static/js/appjs/google_analytics.js"
```
[the remark this answers is not in the document]
Comment (earora97) [reply minify.sh]: Both the files give a parse error on using uglifier, I tried to figure out but still, I am not sure as to why is this happening for these 2 files.
screenshot1
screenshot from 2018-07-21 16-29-30

File: minify.sh [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    "static/js/appjs/layout.js"
+    "static/js/appjs/default/contests.min.js"
```
[the remark this answers is not in the document]
Comment (earora97) [reply minify.sh]: I'll change this. Added the file by mistake.

**Comment by earora97** [comment]
As far as my understanding is concerned, deploy.sh is used to deploy the code to the website. minify.sh is being called in deploy.sh, so why does uglifyjs installation need to be there in the readme? Need user install it?

File: minify.sh [hunk 3]
```diff
@@ -0,0 +1,94 @@
+#!/bin/bash
+    : '
```
[the remark this answers is not in the document]
Comment (earora97) [reply minify.sh]: """ is used in python. This is a bash file. Am I missing something here?

[the remark this answers is not in the document]
Comment (raj454raj) [reply minify.sh]: oh my bad

File: minify.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (raj454raj) [reply minify.sh]: google_analytics.js: https://github.com/stopstalk/stopstalk-deployment/blob/master/static/js/appjs/google_analytics.js#L4 Get rid of the default parameter.
profile.js: https://github.com/stopstalk/stopstalk-deployment/blob/master/static/js/appjs/user/profile.js#L273 You will have remove this kind of anonymous function and add the old conventional `function(er) {return something;}`

File: minify.sh [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    fi
+fi
+
+if [ ${#js_files[@]} != 0 ]
+then
+    for filename in $js_files;
+    do
+        echo "Minifying js file:" $filename
+        newfilename=${filename%$".js"}
+        uglifyjs $filename --output $newfilename.min.js
```
[the remark this answers is not in the document]
Comment (raj454raj) [reply minify.sh]: We also want to mangle the files as the documentation says. This would just remove the extra spaces.

File: views/problems/index.html [hunk 5]
```diff
@@ -2,12 +2,12 @@
 {{block head}}
     <title>{{=(site + " | " + problem_name)}}</title>
     <meta name="description" content="Problem page - {{=(site + ' | ' + problem_name)}}">
-    <link href="{{=get_static_url('css/materialize-tags.css')}}" rel="stylesheet"/>
+    <link href="{{=get_static_url('css/materialize-tags.min.css')}}" rel="stylesheet"/>
```
[the remark this answers is not in the document]
Comment (earora97) [reply views/problems/index.html]: If minified files are used only in the production environment, update_js_timestamp.py should have both the static files and its minified version right?

[the remark this answers is not in the document]
Comment (raj454raj) [reply views/problems/index.html]: Don't think of it that way.

**Production environment** - minified files with versions picked up from redis
**Development environment** - original files without versions. The src tag might as well just have something like `static/js/appjs/user/profile.js`

Both of these functionalities can be handled in `get_static_url` method.
There is no need to run update_js_timestamp.py locally as long as you are not testing the versioning code

**Comment by earora97** [comment]
Few things where I am stuck:
1. You mentioned mangling somewhere in comments. What exactly am I supposed to do in that?
2. How to commit package.json :p
3. I am still unable to figure out changes to be done in profile.js because of which the parse error is coming. I modified the other file though.

ALSO! 
1. If you will check ```get_static_url``` , you'll see that I made changes to return minified file if the environment is production. But there are files which are already minified and the uncompressed version DNE. Should I add an exception for them or unminify them and replace them so that the logic remains same throughout?

**Comment by raj454raj** [comment]
1. This (https://www.npmjs.com/package/uglify-js) has a mangle option. What your uglifyjs does right now is just removes extra spaces to compress the files. We also need to change the variable names to shorter versions to reduce the size of the file further.
2. It's okay if you don't commit package.json for this. package.json is used in npm installations. We will only have one time global installation.
3. Change
```js
years = Array.from(years).map((el)=> parseInt(el));
```
to 
```js
years = Array.from(years).map(function(el) { return parseInt(el); });
```

-----------
1. You are right, there would be a few such occurrences. Let's keep it simple - don't call `get_static_url` on them and remove from the list of files that we use. We don't really need versioning on them as of now.

File: static/js/appjs/google_analytics.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
 (function() {
     'use strict';
 
-    var sendToGA = function(eventCategory, eventLabel, eventAction='click') {
+    var sendToGA = function(eventCategory, eventLabel, eventAction) {
```
Comment (raj454raj) [inline static/js/appjs/google_analytics.js]: 😄 it's not that simple.
Analyze the usage of sendToGA in the file and see where click was used as a default parameter and update the function calls accordingly

Comment (earora97) [reply static/js/appjs/google_analytics.js]: Made all the changes except this one, I will look into it. Until then let me know if there are other changes to be done for this issue :dancer:

File: update_js_timestamp.py [hunk 7]
```diff
@@ -22,46 +22,12 @@
 """
 
 import sys, redis
-import datetime
+import datetime,numpy
```
Comment (raj454raj) [inline update_js_timestamp.py]: is numpy required ?

File: update_js_timestamp.py [hunk 8]
```diff
@@ -122,7 +87,7 @@
     else:
         files = sys.argv[1:]
         if files[0] == "ALL":
-            files = all_js_files
+            files = all_js_css_files+all_images_files
```
Comment (raj454raj) [inline update_js_timestamp.py]: spaces around operator

File: views/default/submissions.html [hunk 9]
```diff
@@ -46,7 +46,7 @@ <h5>Link to the Facebook group: <a target="_blank" style="color: #3B5998;" href=
 </div>
 
 {{if total_rows:}}
-    <script src="{{=get_static_url('js/jquery.bootpag.min.js')}}"></script>
+    <script src="{{=URL('static','js/jquery.bootpag.min.js')}}"></script>
```
Comment (raj454raj) [inline views/default/submissions.html]: space after comma here and at other places

File: files.txt [hunk 10]
```diff
@@ -0,0 +1,35 @@
+js/bloodhound.min.js
```
Comment (raj454raj) [inline files.txt]: file name could be static_files_list.txt

File: minify.sh [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    then
+        static_files=${all_js_css_files[*]}
+    fi
+fi
+
+if [ ${#static_files[@]} != 0 ]
+then
+    for filename in $static_files;
+    do
+        newfilename=static/minified_files/${filename}
```
Comment (raj454raj) [inline minify.sh]: Don't like how files.txt is used. The files should be the list of existing ones and append `.min` wherever required

File: models/db.py [hunk 12]
```diff
@@ -771,10 +771,28 @@ def get_profile_url(site, handle):
...
-             extension=False)
+  if current.environment == "production":
+    new_file_path = file_path
+    if file_path[-3:] == ".js":
+      new_file_path = file_path[:-3] + ".min.js"
+    elif file_path[-4:] == ".css":
+      new_file_path = file_path[:-4] + ".min.css"
+    else:
+      return URL("static",
+               file_path,
```
Comment (raj454raj) [inline models/db.py]: indent file_path under `"static"`

File: static/js/materialize-tags.min.js [hunk 13]
```diff
@@ -0,0 +1 @@
...ue=options[key];options[key]=function(){return value}}}var htmlEncodeContainer=$("<div />");function htmlEncode(value){if(value){return htmlEncodeContainer.text(value).html()}else{return""}}function doGetCaretPosition(oField){var iCaretPos=0;if(document.selection){oField.focus();var oSel=document.selection.createRange();oSel.moveStart("character",-oField.value.length);iCaretPos=oSel.text.length}else if(oField.selectionStart||oField.selectionStart=="0"){iCaretPos=oField.selectionStart}return iCaretPos}function keyCombinationInList(keyDownEvent,lookupList){var found=false;$.each(lookupList,function(index,keyCombination){if(typeof keyCombination==="number"&&keyDownEvent.which===keyCombination){found=true;return false}if(keyDownEvent.which===keyCombination.which){var alt=!keyCombination.hasOwnProperty("altKey")||keyDownEvent.altKey===keyCombination.altKey,shift=!keyCombination.hasOwnProperty("shiftKey")||keyDownEvent.shiftKey===keyCombination.shiftKey,ctrl=!keyCombination.hasOwnProperty("ctrlKey")||keyDownEvent.ctrlKey===keyCombination.ctrlKey;if(alt&&shift&&ctrl){found=true;return false}}});return found}$(function(){$("input[data-role=materialtags]").materialtags()})})(window.jQuery);
```
Comment (raj454raj) [inline static/js/materialize-tags.min.js]: I guess these files were committed earlier but not required in static/js right ?

File: static/flag-icon/css/flag-icon.min.css [hunk 14]
```diff
@@ -1 +0,0 @@
...(../flags/4x3/za.svg)}.flag-icon-za.flag-icon-squared{background-image:url(../flags/1x1/za.svg)}.flag-icon-zm{background-image:url(../flags/4x3/zm.svg)}.flag-icon-zm.flag-icon-squared{background-image:url(../flags/1x1/zm.svg)}.flag-icon-zw{background-image:url(../flags/4x3/zw.svg)}.flag-icon-zw.flag-icon-squared{background-image:url(../flags/1x1/zw.svg)}.flag-icon-eu{background-image:url(../flags/4x3/eu.svg)}.flag-icon-eu.flag-icon-squared{background-image:url(../flags/1x1/eu.svg)}.flag-icon-gb-eng{background-image:url(../flags/4x3/gb-eng.svg)}.flag-icon-gb-eng.flag-icon-squared{background-image:url(../flags/1x1/gb-eng.svg)}.flag-icon-gb-nir{background-image:url(../flags/4x3/gb-nir.svg)}.flag-icon-gb-nir.flag-icon-squared{background-image:url(../flags/1x1/gb-nir.svg)}.flag-icon-gb-sct{background-image:url(../flags/4x3/gb-sct.svg)}.flag-icon-gb-sct.flag-icon-squared{background-image:url(../flags/1x1/gb-sct.svg)}.flag-icon-gb-wls{background-image:url(../flags/4x3/gb-wls.svg)}.flag-icon-gb-wls.flag-icon-squared{background-image:url(../flags/1x1/gb-wls.svg)}.flag-icon-un{background-image:url(../flags/4x3/un.svg)}.flag-icon-un.flag-icon-squared{background-image:url(../flags/1x1/un.svg)}
```
Comment (raj454raj) [inline static/flag-icon/css/flag-icon.min.css]: why is this removed ?

File: static/fa/css/font-awesome.min.css [hunk 15]
```diff
@@ -2,3 +2,4 @@
...
+
```
Comment (raj454raj) [inline static/fa/css/font-awesome.min.css]: avoid unnecessary changes in the PR

File: update_js_timestamp.py [hunk 16]
```diff
@@ -22,46 +22,12 @@
...
-                "materialize/css/materialize.min.css",
-                "css/stopstalk.css",
-                "css/style.css",
-                "css/calendar.css",
-                "css/owlie.css",
-                "fa/css/font-awesome.min.css",
-                "flag-icon/css/flag-icon.min.css",
-                "css/materialize-tags.css",
-                "images/favicon.ico",
+text_file = open("files.txt", "r")
```
Comment (raj454raj) [inline update_js_timestamp.py]: text_file.close() missing

File: update_js_timestamp.py [hunk 17]
```diff
@@ -22,46 +22,12 @@
...
+all_js_css_files = text_file.read().splitlines()
+
+all_images_files = ["images/favicon.ico",
```
Comment (raj454raj) [inline update_js_timestamp.py]: all_image_files

File: static/js/appjs/google_analytics.js (hunk 6 above)
Comment (earora97) [reply static/js/appjs/google_analytics.js]: Made these changes
